pytorch_lightning/utilities/apply_func.py

```diff
@@ -222,6 +222,56 @@ def apply_to_collections(
         ]
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
+    if _is_dataclass_instance(data1):
+        if not _is_dataclass_instance(data2):
+            raise TypeError(
+                "Expected inputs to be dataclasses of the same type or to have identical fields"
+                f" but got input 1 of type {type(data1)} and input 2 of type {type(data2)}"
+            )
+        if not (
+            len(dataclasses.fields(data1)) == len(dataclasses.fields(data2))
+            and map(lambda f1, f2: isinstance(f1, type(f2)), zip(dataclasses.fields(data1), dataclasses.fields(data2)))
+        ):
+            raise TypeError("Dataclasses fields do not match")
+        # make a deepcopy of the data,
+        # but do not deepcopy mapped fields since the computation would
+        # be wasted on values that likely get immediately overwritten
+        data = [data1, data2]
+        fields = [{}, {}]  # type: List[dict]
+        memo = [{}, {}]  # type: List[dict]
+        for i in range(len(data)):
+            for field in dataclasses.fields(data[i]):
+                field_value = getattr(data[i], field.name)
+                fields[i][field.name] = (field_value, field.init)
+                memo[i][id(field_value)] = field_value
+
+        result = deepcopy(data1, memo=memo[0])
+
+        # apply function to each field
+        for ((field_name, (field_value1, field_init1)), (_, (field_value2, field_init2))) in zip(
+            fields[0].items(), fields[1].items()
+        ):
+            if field_init1 and field_init2:
```

Comment:
what if this condition fails, then `v` will not be defined, and the code below might fail.

Reply:
```suggestion
            v = None
            if field_init1 and field_init2:
```

I also applied the same change here in `apply_to_collection`

https://github.com/PyTorchLightning/pytorch-lightning/blob/2060fdec89d7f4585f122d440c252ead9340c519/pytorch_lightning/utilities/apply_func.py#L140-L143